ui/app/selectors/permissions.js

```diff
@@ -194,3 +194,50 @@ function getAccountsCaveatFromPermission (accountsPermission = {}) {
 function domainSelector (state, origin) {
   return origin && state.metamask.domains?.[origin]
 }
+
+export function getAccountToConnectToActiveTab (state) {
+  const selectedAddress = getSelectedAddress(state)
+  const connectedAccounts = getConnectedAccountsForActiveTab(state)
+
+  const { metamask: { identities } } = state
+  const numberOfAccounts = Object.keys(identities).length
+
+  if (connectedAccounts.length && connectedAccounts.length !== numberOfAccounts) {
+    if (connectedAccounts.findIndex(({ address }) => address === selectedAddress) === -1) {
+      return identities[selectedAddress]
+    }
+  }
+
+  return undefined
+}
+
+export function getConnectedAccountsForActiveTab (state) {
+  const { activeTab, metamask } = state
+  const { identities, permissionsHistory } = metamask
+
+  const permissionsHistoryByAccount = permissionsHistory[activeTab.origin]?.['eth_accounts']?.accounts
+
+  return getPermittedAccountsForCurrentTab(state)
+    .map((address) => ({
+      address,
+      name: identities[address].name,
+      lastActive: permissionsHistoryByAccount[address],
+    }))
+    .sort((a, b) => b.lastActive - a.lastActive)
```

Comment:
These should be ordered by `lastSelected` rather than `lastActive`. That's the order they'll be given to the dapp in. The secondary order should be what ever is returned from the Keyring controller.

The implementation for the existing sorting is here: `app/scripts/controllers/permissions/restrictedMethods.js`

Reply:
This is important to get exactly right, so that we don't mistakenly claim the wrong account is primary.

Reply:
Updated to sort on `lastSelected`